### ISSUE TITLE
proposal: refactor MPIR_Init_thread

### ISSUE THREAD

**Issue by hzhou** [pmodels/mpich#3944]
Currently `MPIR_Init_thread` (include `MPID_Init`) is the most complex code because it includes code that touches every components and each components may have a few stages of initialization that depend on different stages of different components. To really comprehend `Init` will need understanding of many parts include those special parts that are only optionally configured in.

The complexity comes from two issues:
* Code from different parts are mixed together out of their respective source sub-tree.
* There are dependency between different parts, and worsened by the fact they are often implicit dependency.

First I would like to propose an idea that will greatly simplify the dependency issue. Let's split `MPIR_init_thread` into three parts: `Init_local`, `Init_node`, and `Init_global`. The pseudo code may look like following:

```
#include "mpir_init.inl"  /* to be addressed */
int MPIR_Init_thread(int *argc, char ***argv, int required, int *provided)
{
    /* code parsing argument and setup thread */
    mpi_errno = MPIR_Init_local();  /* code in Init_local is not aware of ranks or other processes */
    /* code setup POSIX share memory and PMI exchanges */
    mpi_errno = MPIR_Init_node(); /* this stage is world aware, but limited to share memory access */
    mpi_errno = MPIR_Init_global(); /* inter-node exchange, and anything final init/clean up */
}
```

Hopefully, this separation addresses most of the dependency requirement, therefore, may greatly simply code logic. This separation also paves the way for implementing the "session" proposal that is currently going its rounds in the MPI Forum.

To prevent code from spread out of their respective source tree, I propose a simple scheme facilitated by a script just as the way we do with `CVAR`s.

Artificial example for ucx init:
```
/*
=== BEGIN_MPIR_INIT_BLOCK ===
name: ucx_init_1
before: gpu_init
after: ch4_init
[global]
    int ucx_init_local();
[init_local]
    mpi_errno = ucx_init_local();
[init_node]
    MPIR_shm_segs[my_node_id].dev.netmod.ucx.addr = ...;
[Init_global]
    if (my_node_id == 0) {
        MPIR_node_gather(...);
    }
=== END_MPIR_INIT_BLOCK ===
/* 
```
Every parts are optional and each component can have multiple such blocks as fit. The generated `mpir_init.inl` may look like following:
```
/* from src/mpid/ch4/netmod/ucx/ucx_init.c - line xxx */
int ucx_init_local();

int MPIR_Init_local()
{
    int mpi_errno;
    /* from src/mpid/ch4/src/ch4_init.c - line xxx */
    ...
    /* from src/mpid/ch4/netmod/ucx/ucx_init.c - line xxx */
    mpi_errno = ucx_init_local();
  fn_exit:
    return mpi_errno;
}

int MPIR_Init_node()
{
    int mpi_errno;
    /* from src/mpid/ch4/netmod/ucx/ucx_init.c - line xxx */
    MPIR_shm_segs[my_node_id].dev.netmod.ucx.addr = ...;
  fn_exit:
    return mpi_errno;
}

int MPIR_Init_global()
{
    /* from src/mpid/ch4/netmod/ucx/ucx_init.c - line xxx */
    if (my_node_id == 0) {
        MPIR_node_gather(...);
    }
  fn_exit:
    return mpi_errno;
}
```

**Comment by yfguo** [comment]
I have discussed with @hzhou about the refactoring of the initialization. Before we consider splitting the init function into three, here are the refactoring/cleanup items that need to be done:
1. `MPIR_Comm_commit` has special cases for handling the creation of `MPI_COMM_WORLD` that delays calling `MPID_Comm_create_hook` after creating `node_local_comm` and `node_root_comm`. This is because the business card exchange relies on this to make the Allgather collective. As a result, this also ended with the BC exchange happens in `MPID_NM_comm_create_hook` as a special case. A possible fix is to create two functions `MPIR_Comm_init_localcomm` and `MPIR_Comm_init_noderootcomm` that encapsulate their corresponding code that currently live in `MPIR_Comm_commit`. Therefore, we can do this in the `MPID_Init` function without calling `MPIR_Comm_commit`.  Of course, I need to double-check the part that calls `MPID_Comm_init_hook` for the localcomm and node_root_comm. In this way, we probably can tuck the BC exchange function (and address insert etc,) back into the `MPIDI_NM_init_hook`.
2. reorganize the code in `MPID_Init_hook` in a three-stage way: 1) init everything that can be locally initialized (rank, size, nodemap, etc.); 2)SHM mod; 3)netmod.
3. Try to remove the explicit dependency of netmod on PMI.
We discussed more than these, but these are the most actionable items.

**Comment by pavanbalaji** [comment]
This is a good idea and has annoyed me more than once.  Thanks for trying to clean this up.

Please remember that there are two forms of hooks: init hooks and object creation/freeing hooks (e.g., comm objects).  Unfortunately, these are not very well separated, so the init hooks also do some object creation (e.g., for COMM_WORLD).

If object creation/freeing hooks are going to be considered here too, we'll need pre and post hooks for each layer.  So `MPI_Comm_create` will need hooks for `MPID_Comm_create_pre` and `MPID_Comm_create_post` for the device to do work before and after the creation of hooks.  This will get messier when we go one layer further down to netmods, where the same logic would dictate that we'd need `MPIDI_NM_comm_create_pre_pre`, `MPIDI_NM_comm_create_pre_post`, `MPIDI_NM_comm_create_post_pre`, and `MPIDI_NM_comm_create_post_post`.  Luckily that's all the layering we have, so it might not be too bad.  But if there's a more elegant solution, we should certainly consider that.

**Comment by hzhou** [comment]
> Unfortunately, these are not very well separated, so the init hooks also do some object creation (e.g., for COMM_WORLD).

@yfguo and I had a debate with similar category: the dependency of `netmod` on `shmmod`. The dependency is there mostly due to the need of `shm_barrier`. I would like to remove that dependency, which potentially leads to a separate implementation of `posix_shm_init` and init-use-only `shm_barrier`. @yfguo had a very strong argument against this:

* Why maintaining two separate code when one can do?

My reasoning is having two separate (and somewhat duplicate) code decouples the dependency, the benefits of decoupling is much greater than the cost of maintaining the extra piece. First, compared to the full shmmod, maintaining the init-time only posix shm code is trivial, because it has simple well-defined use case, it has zero or trivial dependency, and it does not concern with performance. It probably amounts to less than a hundred line of code and a new developer can maintain it in couple of hours without much knowledge of the rest of the code base. On the other hand, having "shmmod_init" as a dependency is a much heavy dependency. Feature developers who wants to implement feature init using shm need understand the shm run-time. If need new feature, one has to add code into the shmmod. Even one may argue right now it is not complicated, I would argue that the mere existence of potential dependency is real complexity. In addition, in the case of other shmmod such as xpmem, there is also a bootstrapping issue that could be dissolved by having init-only shm utils. Therefore, I think:

* Having separate, init-time only code even with some-degree of functional duplication saves the complexity overall

The object creation/freeing hooks are run-time code. `COMM_WORLD` and `COMM_SELF` are so special that their creation really are init-time only. If possible, it is desirable to make object creation/freeing hooks and init hooks as orthogonal as possible. That may lead to some comm-creation code duplicated for `COMM_WORLD` and `COMM_SELF`. Here I would like to open the can that some duplication (of function) can serve greater good.

**Comment by hzhou** [comment]
@yfguo We have config option `MPIDI_CH4_DIRECT_NETMOD` which switches off the entire SHM mod, another reason that the dependency of netmod on shmmod is undesirable.

**Comment by hzhou** [comment]
Just added `doc/notes/init.txt` to lay out the guide lines and rationale:
     (doc/notes/init.txt)[https://github.com/pmodels/mpich/pull/3956/commits/6d47d4e093343814528d0ec166bc98498e020cd1]
Please feel free to review and comment.

**Comment by hzhou** [comment]
@gcongiu @raffenet When you have a chance, could you take a look at this issue and PR #3956. It is still on-going so it is not really ready for review. But you may just read the notes and randomly glance into some of the commits see if you have any early feedback. Appreciate it!

**Comment by hzhou** [comment]
Hmm... I don't quite like `..._init_final`, it is too similar to `..._finalize()`. `..._init_local` sounds pretty catchy; but both `..._init_node` and `..._init_final` are a bit ambiguous or confusing. Any idea with better names?

I guess we always can do `..._init_1`, `..._init_2`, `..._init_3` or `..._init_stage1`, `..._init_stage2`, and `..._init_stage3`.

**Comment by hzhou** [comment]
Had some discussion with @gcongiu on `MPI_Finalize`. We can't think of a justification that we need call collectives during `MPI_Finalize`. Well, MPI forum defines `MPI_Finalize` as collective, so we need some synchronization at the beginning of `MPI_Finalize` to ensure all processes participates; but after that, every thing seems to be just "local".

**Comment by gcongiu** [comment]
@hzhou `MPI_Finalize` is collective but the standard does not say anything about it being synchronizing (as far as I can tell). MPI implementation just need to make sure that any ongoing communication is completed locally.

**Comment by hzhou** [comment]
I was wondering if we implement `MPI_Finalize` that does not require collective behavior, is it breaking the standard?

**Comment by gcongiu** [comment]
Yes, collective just means it has to be called by all processes. Not all collectives are synchronizing, e.g., `MPI_Bcast` can return before other processes even receive the data if MPI can buffer the bcast data locally. MPI implementation can decide to make it synchronizing but that is not required as far as I have understood.
